narwhals/_duckdb/dataframe.py

```diff
@@ -504,10 +505,21 @@ def unpivot(
             duckdb.sql(query).select(*[*index_, variable_name, value_name])
         )
 
+    def with_row_index(self, name: str, order_by: Sequence[str] | None) -> Self:
+        if order_by is None:
+            msg = (
+                "`LazyFrame.with_row_index` requires `order_by` to be specified as it is an "
+                "order-dependent operation."
+            )
+            raise ValueError(msg)
+        order_by_sql = generate_order_by_sql(*order_by, ascending=True)
+        rel = self.native  # noqa: F841
+        query = f"select row_number() over ({order_by_sql}) -1 as {name}, * from rel"  # noqa: S608
```

Comment:
we may want to use `SQLExpression`, `rel.select`, and `.alias`? this would require duckdb>=1.3, but i think that's ok

we can also use `FunctionExpression('row_number')`